app.ts

```diff
@@ -0,0 +1,216 @@
+            }
+            getRecords(startIndex, startIndex + end);
+        }, time);
+    }
+    
+    // on clicks:
+    searchbtn.onclick = function() {
+        let from = parseInt(fromField.value, 10);
+        // validate field:
+        if (validate(from)) {
+            let end = tableRecordCount - 1;
+            startIndex = from;
+            if (recordCount < startIndex + tableRecordCount) {
+                startIndex = recordCount - tableRecordCount;
+            }
+            getRecords(startIndex, startIndex + end);
+        }
+    };
+
+    nextbtn.onclick = function() {
+        let end = tableRecordCount - 1;
+        if (recordCount > startIndex + tableRecordCount) {
+            startIndex += tableRecordCount;
+            if (recordCount < startIndex + tableRecordCount) {
+                startIndex = recordCount - tableRecordCount;
+            }
+            getRecords(startIndex, startIndex + end);
+        }
+    };
+
+    prevbtn.onclick = function() {
+        if (startIndex >= tableRecordCount) {
+            startIndex -= tableRecordCount;
+        } else if (startIndex < tableRecordCount) {
+            startIndex = 0;
+        }
+        getRecords(startIndex, startIndex + tableRecordCount - 1);
+    };
+
+    $("#canvas").append(fromField,searchbtn, prevbtn, nextbtn);
+
+    initiate();
+
+    /**
+     * Initiates the HTTP requests to obtain the records and column values necessary for the table.
+     */
+    function initiate() {
+        // get number of records:
+        $.ajax({
+            url: "http://localhost:2050/recordCount",
+            success: function (result) {
+                recordCount = parseInt(JSON.parse(result), 10);
+                // get columns:
+                $.ajax({
+                    url: "http://localhost:2050/columns",
+                    success: function (result) {
+                        columns = JSON.parse(result);
+                        numColumns = getSize(columns);
+                        // get first page of records and display them:
+                        adjustTableRecordCount();
+                        getRecords(0, tableRecordCount - 1);
+                    },
+                    error: function (err) {
+                        $("body").text("Error: " + err.status + " " + err.statusText);
+                    }
+                });
+            },
+            error: function (err) {
+                $("body").text("Error: " + err.status + " " + err.statusText);
+            }
+        });
+    }
+
+    /**
+     * Generates a table populated with data, distrubuted evenly.
+     * @param data The data records retrieved from the server, structured as a 2D object array.
+     * @param columns An object array containing the column heading values.
+     */
+    function generateTable(data: any, columns: any) {
+        $("table").remove(); // Remove previous table
+        let dataSize = getSize(data);
```

Comment:
dataSize and perhaps a few other properties here don't change, redeclare as const.